proposals/XXXX-recursive-protocol-constraints.md

```diff
@@ -0,0 +1,265 @@
+* `_RangeReplaceableIndexable`
+
+The following standard library protocols and types will change.
+
+Note that since the specific collection types conform to `Collection`, and `Collection` refines `Sequence`, not all the
+constraints need to be defined on every collection-related associated type.
+
+Default values of all changed associated types remain the same, unless explicitly noted otherwise.
+
+All "Change associated type" entries reflect the complete, final state of the associated type definition, including
+removal of underscored protocols and addition of any new constraints.
+
+### `Any*Collection` (all variants)
+
+* Remove all `C.Subsequence` and `C.Indices` constraints from public `init`
+
+### `Arithmetic`
+
+* Change associated type: `associatedtype Magnitude : Arithmetic`
+
+### `BidirectionalCollection`
+
+* Remove conformance to `_BidirectionalIndexable`
+* Change associated type: `associatedtype SubSequence : BidirectionalCollection`
+* Change associated type: `associatedtype Indices : BidirectionalCollection`
+
+### `Collection`
+
+* Remove conformance to `_Indexable`
+* Change associated type: `associatedtype SubSequence : Collection where SubSequence.Index == Index, SubSequence.Indices == Indices`
+* Change associated type: `associatedtype Indices : Collection where Indices.Iterator.Element == Index, Indices.Index == Index, Indices.SubSequence == Indices`
+
+### `Default*Indices` (all variants)
+
+* Declarations changed to `public struct Default*Indices<Elements : *Collection> : *Collection`
+
+### `IndexingIterator`
+
+* Declaration changed to `public struct IndexingIterator<Elements : Collection> : IteratorProtocol, Sequence`
+
+### `LazyFilter*Collection` (all variants)
+
+* Add default associated type conformance: `typealias SubSequence = ${Self}<Base.SubSequence>`
+
+### `LazyMap*Collection` (all variants)
+
+* Add default associated type conformance: `typealias SubSequence = ${Self}<Base.SubSequence>`
+
+### `Mirror`
+
+* Remove all `C.Subsequence` and `C.Indices` constraints from `init<Subject, C : Collection>(_ subject: Subject, children: C, displayStyle: DisplayStyle?, ancestorRepresentation: AncestorRepresentation)`
+* Remove all `C.Subsequence` and `C.Indices` constraints from `init<Subject, C : Collection>(_ subject: Subject, unlabeledChildren: C, displayStyle: DisplayStyle?, ancestorRepresentation: AncestorRepresentation)`
+
+### `MutableCollection`
+
+* Change associated type: `associatedtype SubSequence : MutableCollection`
+
+### `RandomAccessCollection`
+
+* Change associated type: `associatedtype SubSequence : RandomAccessCollection`
+* Change associated type: `associatedtype Indices : RandomAccessCollection`
+
+### `RangeReplaceableCollection`
+
+* Change associated type: `associatedtype SubSequence : RangeReplaceableCollection`
+
+### `Sequence`
+
+* Change associated type: `associatedtype SubSequence : Sequence where Iterator.Element == SubSequence.Iterator.Element, SubSequence.SubSequence == SubSequence`
```

Comment:
I don't think we need to add the constraint `SubSequence.SubSequence == SubSequence`—this is generally true in practice, but I think we're using it currently as a constraint because it makes the other constraints endlessly recursive. Once the constraints are in place on the protocols we should only need to require that the subsequence is a sequence with the same `Element` type.